src/python/pants/backend/jvm/tasks/resolve_shared.py

```diff
@@ -13,18 +13,39 @@
 class ResolveBase(object):
   """Common methods for both Ivy and Coursier resolves."""
 
-  def add_directory_digests_for_jars(self, jars):
-    """Get DirectoryDigests for jars and return them zipped with the jars.
+  def add_directory_digests_for_jars(self, targets_and_jars):
+    """For each target, get DirectoryDigests for its jars and return them zipped with the jars.
 
-    :param jars: List of pants.java.jar.jar_dependency_utils.ResolveJar
-    :return: List of ResolveJars.
+    :param targets_and_jars: List of tuples of the form (Target, [pants.java.jar.jar_dependency_utils.ResolveJar])
+    :return: list[tuple[(Target, list[pants.java.jar.jar_dependency_utils.ResolveJar])]
     """
+
+    targets_and_jars=list(targets_and_jars)
+
+    if not targets_and_jars:
+      return targets_and_jars
+
+    jar_paths = []
+    for target, jars_to_snapshot in targets_and_jars:
+      [jar_paths.append(fast_relpath(jar.pants_path, get_buildroot())) for jar in jars_to_snapshot]
+
     snapshots = self.context._scheduler.capture_snapshots(
-      tuple(PathGlobsAndRoot(
-        PathGlobs([fast_relpath(jar.pants_path, get_buildroot())]), get_buildroot()) for jar in jars)
-    )
-    return [ResolvedJar(coordinate=jar.coordinate,
-                        cache_path=jar.cache_path,
-                        pants_path=jar.pants_path,
-                        directory_digest=directory_digest) for jar, directory_digest in
-            list(zip(jars, [snapshot.directory_digest for snapshot in snapshots]))]
+      tuple(
+        PathGlobsAndRoot(PathGlobs([jar]), get_buildroot()) for jar in jar_paths
+      ))
+
+    # We want to map back the list[Snapshot] to targets_and_jars
+    # We assume that (1) jars_to_snapshot has the same number of ResolveJars as snapshots does Snapshots,
+    # and that (2) capture_snapshots preserves ordering.
+    digests = [snapshot.directory_digest for snapshot in snapshots]
+    digest_iterator = iter(digests)
```

Comment:
It may be slightly cleaner to use:
`for (target, jars_to_snapshot), digest in zip(targets_and_jars, digests):`

rather than to extract an explicit `iter` and call `next` on it, but probably not sufficiently much to actually change it :)

Reply:
I am not sure of the behavior of that, does it create one ((target, [jars_to_snapshot]), digest) tuple for each digest? If that is the case, it wouldn't help, because we need to map digests to jars, not to (target, [jars]) pairs.

If zip has another behavior, then it could be great! I can look into it tomorrow.